src/base/commands/command.ts

```diff
@@ -52,14 +52,22 @@ export interface ICommand {
 }
 
 /**
- * Commands return the changed model or a Promise for it. The latter
- * serves animating commands to render some intermediate states before
+ * Commands return the changed model or a Promise for it. Promises
+ * serve animating commands to render some intermediate states before
  * finishing. The CommandStack is in charge of chaining these promises,
  * such that they run sequentially only one at a time. Due to that
  * chaining, it is essential that a command does not make any assumption
  * on the state of the model before execute() is called.
  */
-export type CommandResult = SModelRoot | Promise<SModelRoot>;
+export type CommandResult = SModelRoot | Promise<SModelRoot> | DetailedCommandResult;
+
+/**
+ * The `DetailedCommandResult` allows to specify whether the model has changed
+ * and the original action that caused the command to be executed. In case such
+ * an action is given, it is passed to the viewer in order to link any
+ * subsequent response action to the original request.
+ */
+export type DetailedCommandResult = { model: SModelRoot, isChanged: boolean, cause?: Action };
```

Comment:
I'm wondering whether we should rename `isChanged` to `isModelChanged` as it might be unclear in the context of a command what `isChanged` means.

Reply:
Renaming `CommandResult` to `CommandReturn` and `DetailedCommandResult` to `CommandResult`. We may further extend it in the future. Currently we need the `cause` only for request actions, but there is no need to restrict it here since the actual action kind must be checked in the decorators anyway.